ansible: Fix installation when build user != installation user

moreati opened the pull request:
When the ssh user (e.g. pi) did not equal the superuser (e.g. root),
then the ava-install role could not find the binary that had been built.

Using `~{{ ansible_user }}` ensures the path expands to the same
location, regardless of whether privileges have been elevated with
`become: true`.

When these roles were first developed and tested on a DO Droplet the
build user and install user were both root. This meant build and install
happened as the same user.

holisticode commented on scripts/ansible/roles/avalanche_build/defaults/main.yml:
```diff
@@ -1,5 +1,5 @@
-avalanche_binary: ~/go/src/github.com/ava-labs/avalanchego/build/avalanchego
-repo_folder: ~/go/src/github.com/ava-labs/avalanchego
+avalanche_binary: "{{ repo_folder }}/build/avalanchego"
+repo_folder: "~{{ ansible_user }}/go/src/github.com/ava-labs/avalanchego"
```
Where is the `ansible_user` variable set and what value does it have? Is there a default?

I don't see this set anywhere in our codebase, would this need an entry in the `vars/main.yml` file?

moreati commented:
ansible_user is set automatically by A single, to the SSH username you've
connected with, or if the playbook is run locally as the user running it.
In Ansible lingo it's a "special variable"

On Tue, 6 Oct 2020, 16:23 holisticode, <notifications@github.com> wrote:

> *@holisticode* commented on this pull request.
> ------------------------------
>
> In scripts/ansible/roles/avalanche_build/defaults/main.yml
> <https://github.com/ava-labs/avalanchego/pull/291#discussion_r500384633>:
>
> > @@ -1,5 +1,5 @@
> -avalanche_binary: ~/go/src/github.com/ava-labs/avalanchego/build/avalanchego
> -repo_folder <http://github.com/ava-labs/avalanchego/build/avalanchego-repo_folder>: ~/go/src/github.com/ava-labs/avalanchego
> +avalanche_binary: "{{ repo_folder }}/build/avalanchego"
> +repo_folder: "~{{ ansible_user }}/go/src/github.com/ava-labs/avalanchego"
>
> Where is the ansible_user variable set and what value does it have? Is
> there a default?
>
> I don't see this set anywhere in our codebase, would this need an entry in
> the vars/main.yml file?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ava-labs/avalanchego/pull/291#pullrequestreview-503085479>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABKS4S6PAIVDRJ5P35DU73SJMY53ANCNFSM4O2HVZAA>
> .
>

holisticode commented:
That's what I was assuming as well.

But I ran a super dumb test and it was printing 

> "`ansible_user` is undefined"

That's in a supersimple playbook which only does 
```
   - debug:
           msg: "{{ ansible_user }}"
```

moreati commented:
I'm not sure why you got that, doing an adhoc test I get

```
$ ansible -i localhost, -c local localhost -mdebug -avar=ansible_user
localhost | SUCCESS => {
    "ansible_user": "alex"
}
```

with

```
$ ansible --version
ansible 2.9.6
  config file = /home/alex/.ansible.cfg
  configured module search path = ['/home/alex/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3/dist-packages/ansible
  executable location = /usr/bin/ansible
  python version = 3.8.2 (default, Jul 16 2020, 14:00:26) [GCC 9.3.0]
```

holisticode commented:
You are running the `ansible` command while I am trying with `ansible-playbook` - with the latter, `ansible_user` seems not available somehow?

moreati commented:
`ansible-playbook` vs `ansible` doesn't make a difference for me.

1. What version of Ansible are you using?
2. What OS/version are you on?

```
$ cat testme.yml 
- hosts: localhost
  connection: local
  tasks:
    - debug:
        var: ansible_user
    - debug:
        msg: "{{ ansible_user }}"

$ ansible-playbook testme.yml 
[WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost
does not match 'all'

PLAY [localhost] ****************************************************************************************

TASK [Gathering Facts] **********************************************************************************
ok: [localhost]

TASK [debug] ********************************************************************************************
ok: [localhost] => {
    "ansible_user": "alex"
}

TASK [debug] ********************************************************************************************
ok: [localhost] => {
    "msg": "alex"
}

PLAY RECAP **********************************************************************************************
localhost                  : ok=3    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   

$ ansible-playbook --version
ansible-playbook 2.9.6
  config file = /home/alex/.local/go/src/github.com/ava-labs/avalanchego/scripts/ansible/ansible.cfg
  configured module search path = ['/home/alex/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3/dist-packages/ansible
  executable location = /usr/bin/ansible-playbook
  python version = 3.8.2 (default, Jul 16 2020, 14:00:26) [GCC 9.3.0]

holisticode commented:
Sorry for this but I think we are getting closer.

If `connection: local` is not set, or if `remote_user` is not set, then the variable is not known.

````
---
- name: Test the env
  hosts: all

  tasks:
    - debug:
        msg: "User: {{ ansible_user }}"
```

So the above gives me the error with the variable.
Using ansible 2.10.1 BTW.

I am not sure this is a problem at all.
Just want to make sure we are not missing something.

moreati commented:
Why do I have a sense of Deja Vu? *Checks notes*. I'm so sorry, I discovered the same in August, then forgot https://twitter.com/moreati/status/1298014748164919296

> Ansible gotcha:  You're not always who you think you are. In Ansible there are 3 identities you might be running as
> 1. user running the playbook
> 2. user that Ansible connects as (default: same as 1)
> 3. user that Ansible becomes (default: root)

> ansible_user: a variable that reflects the 2nd case. It's often set in an inventory, or specified on the command line, but it doesn't have to be. If the connecting user is not defined in the inventory, then ansible_user can be undefined

holisticode commented:
So my desired end-goal would be I can define the user which I want to install it to explicitly and clearly.

moreati commented:
In all my inventories I've either used `connection: local`; or I've had to set `ansible_user`, because it differed from my local username.

Using `connection: ssh`, without setting `ansible_user` is a corner case. I thought it was quite rare,  but apparently not rare enough.

Unfortunately, I don't know a way to reliably determine "user running this task, when `become == false`" without using ansible_user. As described in those tweets the values gathered in `ansible_facts` only reflect the picture when facts were gathered, they may be different based on whether `become: true` was in effect at the play level or not.

moreati commented:
Yes I think that's doable. The code in this PR would support both local and remote deployment with the following caveat:
- the user must create an inventory
- if the connection is ssh, then `ansible_user` must be explicitly set (either in the inventory, or at the command line)

Here's the inventory (minus IP address) I had when I as running a node on Digital Ocean

```yaml
avalanche_nodes:
  hosts:
    droplet:
      ansible_host: 12.34.56.78
      ansible_user: root
```

to apply the playbook I would run

```
ansible-playbook -i digital_ocean.yml service_playbook.yml
```

to apply it locally I could have the following inventory

```yaml
avalanche_nodes:
  hosts:
    localhost:
      ansible_connection: local
```

and the command would be (assuming sudo requires a password)

```
ansible-playbook -i localhost.yml --ask-become-pass service_playbook.yml
```

moreati commented:
The `localhost.yml` playbook is generic enough that it could be committed to git. Other inventories would probably be too specific/individual to do that.

holisticode commented:
A bit related to this:

Assume we release debian `.deb` packages - or other platform's packages.
The installation would run as `root`.

If we'd like a standard user installation - and not as `root`- what would you suggest, possibly *with* `systemd`?

Looks to me the only sensible approach would be to create a user and launch is as that. But we can't assume `ansible` and thus I would assume an installation script is the way out?

moreati commented:
These Ansible roles (specifically `roles/avalanche_user`) already create a serivce user to run avalanchego (under systemd). The database and logs are owned by that user. The user running the service (default: avalanche) is not the same as the user Ansible connects as (e.g. ubuntu, pi, holisticode).

For system packages (e.g. .deb, .rpm) I think it's a common practice to create a user/group in a pre-install script/hook of the package e.g. https://superuser.com/questions/168461/managing-service-accounts-in-an-rpm-spec.

I totally agree that running the service is root is unnecessary, and highly risky.

moreati commented:
> Should we define `ansible_user` explicitly to avoid corner cases?

No/maybe, depending who you mean by 'we'.

We (the playbook and role authors) should not  set ansible_user - because we don't know the correct value for a user's specific circumstances. 

We (users of the playbook and role) should set ansible_user in our inventory, but your inventory will be specific to you, and mine to me. An end user deploying to Ubuntu on Amazon might want ansible_user=ubuntu, someone deploying to a Digital Ocean droplet would want ansible_user=root, for a raspberry pi it would be ansible_user=pi, etc.

To head off this confusion, I've

1. Switched from `ansible_user`, to `ansible_facts.user_id` (despite being called user_id, it is the username). It has a slightly different tradeoff (detailed in the commit message), that I think fewer people will trip over.
2. Added an inventory for the (probably common) case of running the playbook against localhost.

holisticode reviewed (approved):
I think we may have to cater for two different users here:
  * user is on localhost and wants to setup his own machine to work with an avalanchego binary
  * user sets up an avalanchego binary on a remote host (e.g. AWS EC2 instance) and needs to set it up via ansible and remote ssh

As far as I understand this PR (and I am still not sure I do),
  * the former of the two options is covered. There might be edge cases.
  * the latter has higher variability of options and can not be fully covered with this PR

We should assume that for the second case (and actually for edge cases of the former) a user would be knowledgeable enough to configure this ansible configuration so it works for her case. 

In light that in any case this PR is not critical in any way to the correct operation of the node, and also being appreciative of the very welcomed contribution from the community, I will approve this PR.

If you have any further comment to this review, which may help clarify the above, please add it nevertheless to this PR.

Thank you!

moreati commented:
> I think we may have to cater for two different users here:
> 
> * user is on localhost and wants to setup his own machine to work with an avalanchego binary
> * user sets up an avalanchego binary on a remote host (e.g. AWS EC2 instance) and needs to set it up via ansible and remote ssh
> 
> As far as I understand this PR (and I am still not sure I do),
> 
> * the former of the two options is covered. There might be edge cases.
> * the latter has higher variability of options and can not be fully covered with this PR

100% correct.

> P.S. It needs a new rebase

Doing so now

moreati commented:
This change, and the other Ansible changes are collected in https://github.com/moreati/avalanchego/tree/ansible-rollup, should you prefer that.

moreati commented:
Triggering rebuild

moreati commented:
As before I've collected this and the other open Ansible PRs in [moreati:ansible-rollup](https://github.com/moreati/avalanchego/tree/ansible-rollup).

moreati commented:
As before I've collected this and the other open Ansible PRs in [moreati:ansible-rollup](https://github.com/moreati/avalanchego/tree/ansible-rollup).
